sound/soc/intel/boards/sof_sdw.c

```diff
@@ -1277,6 +1278,34 @@ static int sof_sdw_rtd_init(struct snd_soc_pcm_runtime *rtd)
 		 */
 		if (codec_info->dais[dai_index].rtd_init_done)
 			continue;
+
+		/*
+		 * Add card controls and dapm widgets for the first codec dai.
+		 * The controls and widgets will be used for all codec dais.
+		 */
+		if (i == 0) {
+			const struct snd_soc_dapm_widget *w = codec_info->dais[dai_index].widgets;
+			const struct snd_kcontrol_new *c = codec_info->dais[dai_index].controls;
+			int num_c = codec_info->dais[dai_index].num_controls;
+			int num_w = codec_info->dais[dai_index].num_widgets;
+
+			if (c) {
+				ret = snd_soc_add_card_controls(card, c, num_c);
+				if (ret) {
+					dev_err(card->dev, "%#x controls addition failed: %d\n",
+						codec_info->part_id, ret);
+					return ret;
+				}
+			}
+			if (w) {
+				ret = snd_soc_dapm_new_controls(&card->dapm, w, num_w);
+				if (ret) {
+					dev_err(card->dev, "%#x widgets addition failed: %d\n",
+						codec_info->part_id, ret);
+					return ret;
+				}
+			}
+		}
```

Comment:
> could we move all this to line 1272? This is a bit orthogonal to the rtd stuff and now it's in the middle. It'd be cleaner if the two parts didn't interfere.

No, we can't add the controls/widgets when rtd_init_done is true. Otherwise, it will be done twice for both playback and capture links. Those additions were done in `codec_info->dais[dai_index].rtd_init(rtd)`, we just move them outside the .rtd_init, but don't want to change the behavior.

Reply:
Arguably you could also save the snd_soc_dapm_route for like with controls/widgets and do the snd_soc_dapm_add_routes() in this level.

Reply:
> Arguably you could also save the snd_soc_dapm_route for like with controls/widgets and do the snd_soc_dapm_add_routes() in this level.

Right, but routes are very codec specific. So, I prefer keep it in the .rtd_init callback.

Reply:
The widgets and controls also codec specific in some cases...

Reply:
> The widgets and controls also codec specific in some cases...

The controls/widgets are at card level, and we can use our own name. But, routes need to map to the input/output widgets in codec drivers. Besides, we only want to add the controls/widgets once in a link, but we need to add the routes for each codec dai.